### ISSUE TITLE
Significant quality reduction in images when scaling on Android

### ISSUE THREAD

**Issue by dpolivy** [apache/cordova-plugin-camera#600]
# Bug Report

### What is expected to happen?

When setting `targetWidth` and `targetHeight` on the captured image, one expects to get the highest quality scaled image possible for those target dimensions.

### What does actually happen?

When starting with a high resolution image (e.g. from a Pixel 2 or higher), if the target size is relatively small, when loading the unscaled image in to scale it down, a sample size is set on the `BitmapFactory`:

`options.inSampleSize = calculateSampleSize(rotatedWidth, rotatedHeight,  widthHeight[0], widthHeight[1]);`
[code ref](https://github.com/apache/cordova-plugin-camera/blob/master/src/android/CameraLauncher.java#L1048)

Later in the code, the Bitmap is scaled using this code:

`Bitmap scaledBitmap = Bitmap.createScaledBitmap(unscaledBitmap, scaledWidth, scaledHeight, true);`

The problem here is that the first method of scaling the bitmap by sampling pixels leads to a very pixelated image with poor quality. The second method actually yields quite good results in scaling. The difference between the two approaches yields significant differences in the usability of the resulting images.

Here are the camera settings used:
````
		sourceType: Camera.PictureSourceType.CAMERA,
		quality: 75,
		targetWidth: 620,
		targetHeight: 620,
		destinationType: Camera.DestinationType.FILE_URI,
		mediaType: Camera.MediaType.PICTURE,
		correctOrientation: true
````

These images were captured with a Pixel 2 using the above settings.

**Image 1: Captured with the default behavior of calculating a sample size. Notice the fuziness/pixelation around the text**
![image](https://user-images.githubusercontent.com/3200655/82268748-9d96ce80-9924-11ea-94fd-431d68dd4eaa.png)

**Image 2: Captured with the setting of `options.inSampleSize` commented out, and only scaled with `Bitmap.createScaledBitmap`. The text is much clearer overall.**
![image](https://user-images.githubusercontent.com/3200655/82268804-b7d0ac80-9924-11ea-89f9-8db26839370d.png)

## Information

This code to set the sample size appears to be part of the plugin from the beginning. Perhaps it was a useful memory optimization, and the quality degradation was not as noticeable when camera resolutions were much lower. However, with higher resolution cameras on modern devices, if requesting a scaled image, the chance for getting a poorly scaled result greatly increases using this method. It seems better to only use `Bitmap.createScaledBitmap` to do the scaling, and hope the added memory usage of reading in the uncompressed bitmap isn't an issue on modern devices.

### Command or Code

Pixel 2+ with the following options:
````
		sourceType: Camera.PictureSourceType.CAMERA,
		quality: 75,
		targetWidth: 620,
		targetHeight: 620,
		destinationType: Camera.DestinationType.FILE_URI,
		mediaType: Camera.MediaType.PICTURE,
		correctOrientation: true
````

### Environment, Platform, Device

Tested with Pixel 2, Android 10, latest security updates. I also have users reporting similar issues with a Pixel 4, and Samsung Galaxy S9, both also on Android 10.

### Version information
cordova-android: 7.1.1
cordova-plugin-camera: 4.2.0

## Checklist

- [x] I searched for existing GitHub issues
- [ ] I updated all Cordova tooling to most recent version
- [x] I included all the necessary information above

**Comment by dpolivy** [comment]
After doing a lot of experimentation, I think there are a few areas for improvement in how the plugin manages scaling. In my scenario above, with a Pixel 2 at a resolution of 4032 x 3024, to scale down to 620px max is yielding a sample size of 6. So Android is only using every 6th pixel of the original image, and clearly that's not producing a great result.

When I adjust my maximum dimensions to achieve a sample size of 4 or lower, then the scaling result is actually quite good quality.

I understand that the original scaling code here may have been Android best practice in the past, however I think it can be improved upon. I found a [great overview of scaling options on Android](https://stackoverflow.com/a/32121059/1347600) that uses a combination of the `inSampleSize`, `inScaled`, `inDensity` and `inTargetDensity` options to scale the image directly to the desired size in a single API call. Using this approach would eliminate the need to call `Bitmap.createScaledBitmap` again after loading the image. Additionally, since this plugin is dealing with at most one image at a time (vs loading lots of potentially large Bitmaps at once), it seems that maybe it can be a bit more open to using more memory upfront to achieve a better end result.

Secondly, I might suggest that we limit the sample size to a maximum of 4, which would help save some memory on initial load, but subsequently provide a better starting image for the actual scaling process to act upon. There are some issues in the way `calculateSampleSize` works: the documentation says:

> Note: the decoder uses a final value based on powers of 2, any other value will be rounded down to the nearest power of 2.

The function doesn't adhere to that, and seems to return various values that are not powers of 2. In reality, it seems Android itself [may support this](https://stackoverflow.com/questions/37596945/why-bitmapfactory-options-insamplesize-doesnt-be-rounded-down-to-the-nearest-po), although it may be less efficient for the codec, and/or it could lead to some of the quality issues I'm seeing.

I suppose the ultimately resolution here depends on what the goals are of the scaling. In my opinion, it should be the highest quality representation possible for the desired resolution. If that means slightly higher memory use for a short period while the image is loaded and scaled, that might be worth it. Heck, even loading the full image and scaling it down properly may use less memory by only loading the image once, vs the current approach which loads it 2 or 3 times (depending also on whether it is to be rotated).

For now, I've worked around this issue by increasing our maximum dimensions so the scaling result is better, but it does seem like this could be improved in the plugin.

Anyone else have thoughts on the best behavior? I'm open to try and put together a PR if there is some interest and discussion.

**Comment by breautek** [comment]
I think a demo app might be worth while here so that we can see the before and after, and could serve as a testing app for variety of different devices.

PRs are much welcome, but I'm not sure what would be the best path to take. My concern is mostly with budget phones that are maybe run on limited memory... but at the same time I'm not sure if they are even as significant as they were in the past.

**Comment by D-Marc1** [comment]
The quality degrades pretty heavily for me on my iPhone SE 2020 as well, even with `quality: 100`.

**Comment by dpolivy** [comment]
@D-Marc1 The `quality` parameter is related to the jpeg compression of the final image. If you're seeing significant degradation at `100` then it could be a scaling issue as well. What are your values for `targetWidth` and `targetHeight`?

**Comment by D-Marc1** [comment]
@dpolivy Thanks for clearing that up. I guess I misread that in the docs.

As far as `targetWidth` and `targetHeight` I actually didn't set either in my initial tests. In the docs, it doesn't seem to say it's required, but it seems like it might be, as setting the `targetWidth` and `targetHeight` to 3840 x 2160 seemed to fix it for me. Maybe I'll send a pull request and edit the README to specify that you need to explicitly use a width and height, otherwise it'll come out blurry.

**Comment by dpolivy** [comment]
@D-Marc1 Odd, they shouldn't be required. In looking at the iOS code, not setting `targetWidth` and `targetHeight` should cause the image to be full size. You might want to open a separate issue about this and post some comparison images, since the iOS code is separate from Android.

**Comment by sardapv** [comment]
I have a similar issue, text in the image is getting blurred as per the customer. And the case here its a 'cheque' image so MICR number is not readable -_-. Can anyone suggest me ideal height and width, please? If I don't set targetWidth and targetHeight, the image that I am storing in the InAppBrowser's localStorage crossed threshold and throws out of memory error (image is of 1.4MB, and 1900x600, still  base64 encoded is above 25mb!

**Comment by sardapv** [comment]
Can you please tell me did you just comment out line 1081 from code ref you mentioned? what changes I have to for achieving the result same as your 2nd mug pic? Thanks, its very urgent @dpolivy. Also will this be applicable for iOS?

**Comment by dpolivy** [comment]
@sardapv Yes, commenting out that line should help, potentially at the expense of higher memory usage, though. This specific issue is only on Android, the iOS scaling generally seems to give higher quality results.

**Comment by sardapv** [comment]
okay thanks @dpolivy  I will try this. For now i removed height and width and its working fine on android with good quality. but app crashed badly on iOS when storing base64 into localStorage because of memory issue :(. I will give a try to your suggestion.
